src/compiler/checker.ts

```diff
@@ -28040,6 +28036,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getUnionType(map(clauseWitnesses, text => text ? narrowTypeByTypeName(type, text) : neverType));
         }
 
+        function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+            // Equal start and end denotes implicit fallthrough; undefined marks explicit default clause.
+            const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);
+            const hasDefaultClause = clauseStart === clauseEnd || (defaultIndex >= clauseStart && defaultIndex < clauseEnd);
+            if (hasDefaultClause) {
+                // Here we'd need to find the opposite of all of the other cases, but without control flow,
+                // the best we can do is to not narrow at all.
+                return type;
+            }
+            // In the non-default cause we create a union of the type narrowed by each of the listed cases.
+            const clauses = switchStatement.caseBlock.clauses.slice(clauseStart, clauseEnd);
+            return getUnionType(map(clauses, clause => clause && clause.kind === SyntaxKind.CaseClause ? narrowType(type, clause.expression, /*assumeTrue*/ true) : neverType));

```

Comment:
When would `clause` not be truthy?

Reply:
Never, unless we supported `switch (false)`

Reply:
I don't understand, these are each syntax nodes, right?

Reply:
Oh, you're talking about `clause && clause.kind` and not `assumeTrue`.

Yes, I can undo this; I just copied and pasted from the original code which could have had it be falsy.